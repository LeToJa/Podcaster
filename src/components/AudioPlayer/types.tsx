export interface AudioPlayerTypes {
    url: string
    type: string
}