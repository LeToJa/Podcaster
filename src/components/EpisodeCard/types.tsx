export interface EpisodeCardTypes {
    podcastId: string
    episodeId: number
    title: string
    date: string
    duration: string
}