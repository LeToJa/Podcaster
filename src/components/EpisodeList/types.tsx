import { RSSResponse } from '../../helpers/RSSParser/types'

export interface EpisodeListTypes {
    podcastId: string
    episodes: RSSResponse[] | false
}