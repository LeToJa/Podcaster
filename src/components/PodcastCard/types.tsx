export interface PocastCardTypes {
    id: string
    name: string
}