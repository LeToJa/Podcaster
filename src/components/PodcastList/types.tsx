import { PocastCardTypes } from '../PodcastCard/types'

export interface PodcastListTypes {
    podcasts: PocastCardTypes[]
}