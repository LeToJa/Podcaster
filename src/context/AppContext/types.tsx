export interface AppContextTypes {
    loading: boolean
    toggleLoading: (value: boolean) => void
}