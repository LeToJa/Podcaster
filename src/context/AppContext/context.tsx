import { createContext, useState } from 'react'

import { ReactElementWithChildrenTypes } from '../../types'
import { AppContextTypes } from './types'

export const AppContext = createContext<AppContextTypes | undefined>(undefined)

export const AppProvider = ({ children }: ReactElementWithChildrenTypes) => {
    const [loading, setLoading] = useState<boolean>(true)

    const toggleLoading = (value: boolean) => setLoading(value)

    return <>
        <AppContext.Provider value={{ loading, toggleLoading }}>
            {children}
        </AppContext.Provider>
    </>
}