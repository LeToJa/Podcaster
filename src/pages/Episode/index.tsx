import { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'

import { useApp } from '../../context/AppContext'

import { getPodcastInfo } from '../../helpers/ItunesAPI'

import PodcastLayout from '../../layout/PodcastLayout'
import EpisodePanel from '../../components/EpisodePanel'

import { PodcastAsideTypes } from '../../layout/PodcastLayout/types'
import { RSSResponse } from '../../helpers/RSSParser/types'

const Episode = () => {
    const { podcastId, episodeId } = useParams()
    const { loading, toggleLoading } = useApp()
    const [aside, setAside] = useState<PodcastAsideTypes | false>(false)
    const [episode, setEpisode] = useState<RSSResponse | false>(false)

    useEffect(() => {
        const fetchPodcasts = async () => {
            const content = await getPodcastInfo(String(podcastId))

            if (!content) {
                setAside(false)
                toggleLoading(false)
                return
            }

            setAside({
                id: content.id,
                title: content.title,
                author: content.author,
                artwork: content.artwork,
                description: content.description,
            })

            if (!content.episodes || typeof episodeId === 'undefined') {
                toggleLoading(false)
                return
            }

            setEpisode(content.episodes[episodeId as unknown as number])

            toggleLoading(false)
        }

        toggleLoading(true)
        fetchPodcasts().catch(err => console.error(err))
    }, [podcastId])

    if (loading) return

    if (!aside || !episode) return <section className='shadow-xl p-4'>This episode doesn&apos;t work</section>

    return <>
        <PodcastLayout aside={aside}>
            <EpisodePanel {...episode} />
        </PodcastLayout>
    </>
}

export default Episode