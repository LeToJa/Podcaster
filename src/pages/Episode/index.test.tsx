import { render, waitFor } from '@testing-library/react'
import { describe, it, expect, vi } from 'vitest'
import { BrowserRouter, Params } from 'react-router-dom'

import { AppContextProps } from '../../context/AppContext/types'
import { GenericObject, ReactElementWithChildrenTypes } from '../../types'
import { AppContext } from '../../context/AppContext/context'
import { getPodcastInfo } from '../../helpers/ItunesAPI'

import Episode from '.'

vi.mock('react-router-dom', async (importOriginal): Promise<unknown> => {
    const actual: GenericObject = await importOriginal()

    return {
        ...actual,
        useParams: (): Readonly<Params<string>> => ({ podcastId: '123', episodeId: '0' }),
    }
})

vi.mock('../../helpers/ItunesAPI', () => ({
    getPodcastInfo: vi.fn(),
}))

const mockAppContext = {
    loading: false,
    toggleLoading: vi.fn()
}

const MockAppProvider = ({ loading, toggleLoading, children }: AppContextProps & ReactElementWithChildrenTypes) => {
    return (
        <AppContext.Provider value={{ loading, toggleLoading }}>
            {children}
        </AppContext.Provider>
    )
}

const mockPodcast = {
    id: '123',
    title: 'Sample Podcast',
    author: 'Author Name',
    artwork: 'some-url.jpg',
    description: 'Podcast Description',
    episodes: [
        {
            title: 'Episode 1',
            content: '<p>Episode content in HTML format</p>',
            isoDate: '2023-08-29T12:00:00Z',
            enclosure: {
                url: 'http://example.com/audio.mp3',
                type: 'audio/mpeg',
            },
            itunes: {
                duration: '30:00',
            }
        }
    ],
}

const mockGetPodcastInfo = vi.mocked(getPodcastInfo)

describe('Episode', () => {
    it('renders correctly and matches snapshot', async () => {
        mockGetPodcastInfo.mockResolvedValue(mockPodcast)

        const { container } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <Episode />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(container).toMatchSnapshot()
        })
    })

    it('renders "This episode doesnt work" when there is no podcast data', async () => {
        mockGetPodcastInfo.mockResolvedValueOnce(false)

        const { getByText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <Episode />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText(/This episode doesn't work/i)).toBeTruthy()
        })
    })

    it('renders "This episode doesnt work" when there is no episode data', async () => {
        mockGetPodcastInfo.mockResolvedValueOnce({
            ...mockPodcast,
            episodes: false,
        })

        const { getByText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <Episode />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText(/This episode doesn't work/i)).toBeTruthy()
        })
    })

    it('renders podcast layout and episode panel when data is available', async () => {
        mockGetPodcastInfo.mockResolvedValue(mockPodcast)

        const { getByText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <Episode />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText(/Episode 1/i)).toBeTruthy()
            expect(getByText(/Episode content in HTML format/i)).toBeTruthy()
        })
    })
})
