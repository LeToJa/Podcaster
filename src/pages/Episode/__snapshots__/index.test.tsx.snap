// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Episode > renders correctly and matches snapshot 1`] = `
<div>
  <section
    class="shadow-xl p-4"
  >
    This episode doesn't work
  </section>
</div>
`;
